src/universal-dapp.js

```diff
@@ -461,103 +460,103 @@ UniversalDApp.prototype.pendingTransactions = function () {
   return this.txRunner.pendingTxs
 }
 
-UniversalDApp.prototype.runTx = function (args, cb) {
-  var self = this
-  var tx = {
-    to: args.to,
-    data: args.data,
-    useCall: args.useCall
-  }
-  async.waterfall([
-    // query gas limit
-    function (callback) {
-      tx.gasLimit = 3000000
-
-      if (self.transactionContextAPI.getGasLimit) {
-        self.transactionContextAPI.getGasLimit(function (err, ret) {
-          if (err) {
-            return callback(err)
-          }
-
-          tx.gasLimit = ret
-          callback()
-        })
-      } else {
-        callback()
-      }
-    },
-    // query value
-    function (callback) {
-      tx.value = 0
-      if (tx.useCall) return callback()
-      if (self.transactionContextAPI.getValue) {
-        self.transactionContextAPI.getValue(function (err, ret) {
-          if (err) {
-            return callback(err)
-          }
-
-          tx.value = ret
-          callback()
-        })
-      } else {
-        callback()
-      }
-    },
-    // query address
-    function (callback) {
-      if (self.transactionContextAPI.getAddress) {
-        self.transactionContextAPI.getAddress(function (err, ret) {
-          if (err) {
-            return callback(err)
-          }
-
-          tx.from = ret
+function execute (pipeline, env, callback) {
+  function next (err, env) {
+    if (err) return callback(err)
+    var step = pipeline.shift()
+    if (step) step(env, next)
+    else callback(null, env.result)
+  }
+  next(null, env)
+}
 
-          callback()
-        })
-      } else {
-        self.getAccounts(function (err, ret) {
-          if (err) {
-            return callback(err)
-          }
+UniversalDApp.prototype.replayTx = function (args, cb) {
+  var self = this
+  self.getAccounts(function (err, accounts = []) {
+    if (err) console.error(err)
+    if (args.to && args.to[0] === '<') args.to = accounts[args.to.split('>')[0].slice(11)]
```

Comment:
I think that part of the job should be done in the recorder

Reply:
How do you mean that?

Reply:
I thought we will even give the user an option to set custom addresses.
So we don't want to retrieve information like `<account - 1>` from the recorder?

Reply:
the recorder takes care of replacing real address to <account - n> , so i guess this is also his job to convert the other way around (from <account - n> to real address).
I think the `universal-dapp` does not have to know about anything related to these translations.
the udapp just takes a tx that is quasi filled with all the needed param (except gasLimit).

For setting custom addresses yes we can do that as well in the recorder or another modules yet to be coded, but not in udapp.

Reply:
I kind of agree, but then... look at the code (from `run-tab.js`)
```js
// ...
/* ------------------------------------------------
    RECORDER
------------------------------------------------ */
function makeRecorder (appAPI, appEvents) {
  var udapp = appAPI.udapp()
  var recorder = new Recorder({
    events: {
      udapp: appEvents.udapp,
      executioncontext: executionContext.event
    },
    api: {
      getAccounts (cb) { udapp.getAccounts(cb) }
    }
  })

// ...

var runButton = yo`<button class=${css.runTxs}>re-run transactions</button>`

// ....

  runButton.onclick = () => {
    var opts = { title: `Enter Transactions`, text: `Paste the array of transaction you want to replay here`, inputValue: '', multiline: true }
    modalDialogCustom.promptMulti(opts, function confirm (json = '[]') {
      try {
        var txArray = JSON.parse(json)
      } catch (e) {
        modalDialogCustom.alert('Invalid JSON, please try again')
      }
      if (txArray.length) {
        txArray.forEach(tx => udapp.rerunTx(tx.record, CALLBACK))
      }
    }, function cancel () { })
  }
```

Reply:
I think this somehow needs to be re-architected.

1. `run-tab.js` uses `udapp` and a `recorder`, ... which depends on `udapp` as a "passed in/injected" dependency.

2. now `run-tab.js` has of course the buttons to copy and re-run transactions and the event listeners use `udapp.rerunTx` - which currently makes sense, because we are re-using A LOT of code from `UniversalDApp.prototype.runTx` ... because `.replayTx(...)` or now `.rerunTx(...)` is very similar.

**now what you are saying is**
The `recorder` in runtab.js, which already get's `udapp events` injected...
...would need to be accessible inside of `udapp` for the `.rerunTx` method?
Or i would move that method to recorder instead, but it heavily re-uses udapp.runTx stuff...

Somehow everything is using everything else and it feels to me essentially like spagetthi code.

I could probably think about it and come up with some kind of a little bit bigger refactoring to untangle this, but maybe you have a good suggestion how to deal with this situation.

Reply:
I am not saying that the recorder should be accessible inside udapp.
I am saying that `the udapp just takes a tx that is quasi filled with all the needed param (except gasLimit).`

The process of converting `<account - n>` to real addresses should not be done inside `rerunTx` but should be done just before calling `rerunTx` and should be done inside `recorder`.
Actually the process of translating addresses should perhaps be done in an another module, but keeping the all the code in the `recorder` is fine for now.

btw the name `rerunTx` is not really appropriate. Because from the point of view of `udapp`, we just *run* a transaction (well I can't right now figure out a name, but will think on that).